kv.go

```diff
@@ -0,0 +1,436 @@
+	history := int64(1)
+	if cfg.History > 0 {
+		history = int64(cfg.History)
+	}
+	replicas := cfg.Replicas
+	if replicas == 0 {
+		replicas = 1
+	}
+
+	scfg := &StreamConfig{
+		Name:              fmt.Sprintf(kvBucketNameTmpl, cfg.Bucket),
+		Description:       cfg.Description,
+		Subjects:          []string{fmt.Sprintf(kvSubjectsTmpl, cfg.Bucket)},
+		MaxMsgsPerSubject: history,
+		MaxBytes:          cfg.MaxBytes,
+		MaxAge:            cfg.TTL,
+		MaxMsgSize:        cfg.MaxValueSize,
+		Storage:           cfg.Storage,
+		Replicas:          replicas,
+	}
+
+	if _, err := js.AddStream(scfg); err != nil {
+		return nil, err
+	}
+
+	kv := &kvs{
+		name:   cfg.Bucket,
+		stream: scfg.Name,
+		pre:    fmt.Sprintf(kvSubjectsPreTmpl, cfg.Bucket),
+		js:     js,
+	}
+	return kv, nil
+}
+
+// DeleteKeyValue will delete this KeyValue store (JetStream stream).
+func (js *js) DeleteKeyValue(bucket string) error {
+	stream := fmt.Sprintf(kvBucketNameTmpl, bucket)
+	return js.DeleteStream(stream)
+}
+
+type kvs struct {
+	name   string
+	stream string
+	pre    string
+	js     *js
+}
+
+// Underlying entry.
+type kve struct {
+	bucket   string
+	key      string
+	value    []byte
+	revision uint64
+	delta    uint64
+	created  time.Time
+	op       KeyValueOp
+}
+
+func (e *kve) Bucket() string        { return e.bucket }
+func (e *kve) Key() string           { return e.key }
+func (e *kve) Value() []byte         { return e.value }
+func (e *kve) Revision() uint64      { return e.revision }
+func (e *kve) Created() time.Time    { return e.created }
+func (e *kve) Delta() uint64         { return e.delta }
+func (e *kve) Operation() KeyValueOp { return e.op }
+
+// Get returns the latest value for the key.
+func (kv *kvs) Get(key string) (KeyValueEntry, error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	m, err := kv.js.GetLastMsg(kv.stream, b.String())
+	if err != nil {
+		return nil, err
+	}
+
+	entry := &kve{
+		bucket:   kv.name,
+		key:      key,
+		value:    m.Data,
+		revision: m.Sequence,
+		created:  m.Time,
+	}
+
+	// Double check here that this is not a DEL Operation marker.
+	if len(m.Header) > 0 && m.Header.Get(kvop) == kvdel {
+		entry.op = KeyValueDelete
+		return entry, ErrMSgNotFound
+	}
+
+	return entry, nil
+}
+
+// Put will place the new value for the key into the store.
+func (kv *kvs) Put(key string, value []byte) (revision uint64, err error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	pa, err := kv.js.Publish(b.String(), value)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Create will add the key/value pair iff it does not exist.
+func (kv *kvs) Create(key string, value []byte) (revision uint64, err error) {
+	v, err := kv.Update(key, value, 0)
+	if err == nil {
+		return v, nil
+	}
+	// TODO(dlc) - Since we have tombstones for DEL ops for watchers, this could be from that
+	// so we need to double check.
+	if e, err := kv.Get(key); err == ErrMSgNotFound {
```

Comment:
will this code ever actually be reached? since the error handling above wold bail if the first update failed, we'll never get a chance to correct that to be a update with sequence

Reply:
Will take a look.